### PR TITLE
e2e: add test scheduler picks TM policy update without need to restart

### PR DESCRIPTION
**Pull request by rshemtov13**
This test add automation coverage for OCPBUGS-34583

The numaresources secondary-scehduler doesn't account for topology-manager new configuration unless it gets deleted and restarted again. This means TAE can still occur if topology manager policy was updated to single-numa-node and the scheduler not restarted after that update.
This test checks that scenario by ensuring the TM policy is set to none then restarting the scheduler, setting back the TM policy to single-numa-node and then creating a TAE deployment to check if the deployment is pending

### CODE REVIEW

File: test/e2e/serial/tests/configuration.go [hunk 1]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
 				return kcCmNamesCur
 			}).WithContext(ctx).WithTimeout(5 * time.Minute).WithPolling(10 * time.Second).Should(Equal(kcCmNamesPre))
 		})
+
+		It("[reboot_required][slow] should be able to correctly identify topology manager policy without scheduler restarting", Label("reboot_required", "slow"), func() {
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:701]: let's also add tags/labels of unsched and schedrst

File: test/e2e/serial/tests/configuration.go [hunk 2]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			nroOperObj := &nropv1.NUMAResourcesOperator{}
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:702]: let's add a By() step here for fetching the matching kubeletconfig for the numaresources-operator CR

File: test/e2e/serial/tests/configuration.go [hunk 3]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+					}
+				}
+			}
+			Expect(targetedKC).ToNot(BeNil(), "there should be at least one kubeletconfig.machineconfiguration object")
+
+			kcObj, err := kubeletconfig.MCOKubeletConfToKubeletConf(targetedKC)
+			Expect(err).ToNot(HaveOccurred())
+			initialTopologyManagerPolicy := kcObj.TopologyManagerPolicy
+
+			By("modifying Topology Manager Policy under kubeletconfig to none")
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:732]: just a side note, no action needed now, the kubeletconfig can be owned by a performanceprofile (d/s we support both) this need to be adjusted to detect on which object it should be done. I'll post a PR soon for this

File: test/e2e/serial/tests/configuration.go [hunk 4]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+				applyNewTopologyManagerPolicy(&kcObj.TopologyManagerPolicy, v1beta1.NoneTopologyManagerPolicy)
+				err = kubeletconfig.KubeletConfToMCKubeletConf(kcObj, targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				err = fxt.Client.Update(context.TODO(), targetedKC)
+				g.Expect(err).ToNot(HaveOccurred())
+			}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
+			waitForMcpsCondition(fxt.Client, context.TODO(), mcps, machineconfigv1.MachineConfigPoolUpdated)
+
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:748]: after any kubeletconfig update the nodes reboot so need to wait for mcp to get updated, basically similar to what you did in lines 773-778.

File: test/e2e/serial/tests/configuration.go [hunk 5]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			defer func() {
+				By("reverting kubeletconfig changes")
+				kcObj.TopologyManagerPolicy = initialTopologyManagerPolicy
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:751]: not sure why is this needed as kcObj is overridden in 756

File: test/e2e/serial/tests/configuration.go [hunk 6]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+				By("waiting for mcp to start updating")
+				waitForMcpsCondition(fxt.Client, context.TODO(), mcps, machineconfigv1.MachineConfigPoolUpdating)
+
+				By("wait for mcp to get updated")
+				waitForMcpsCondition(fxt.Client, context.TODO(), mcps, machineconfigv1.MachineConfigPoolUpdated)
+			}()
+
+			var schedulerName string
+			nroSchedObj := &nropv1.NUMAResourcesScheduler{}
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:781]: defining the type `var nroSchedObj &nropv1.NUMAResourcesSchedule` can be also be enough

File: test/e2e/serial/tests/configuration.go [hunk 7]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			schedPods, err := podlist.With(fxt.Client).ByDeployment(context.TODO(), *schedDp)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(len(schedPods)).To(Equal(1))
+
+			schedulerName = nroSchedObj.Status.SchedulerName
+			Expect(schedulerName).ToNot(BeEmpty(), "cannot autodetect the TAS scheduler name from the cluster")
+
+			// restart the secondary-scheduler pod
+			By(fmt.Sprintf("deleting the NRO Scheduler object: %s", nroSchedObj.Name))
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:798]: the test is actually deleting the pod

File: test/e2e/serial/tests/configuration.go [hunk 8]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			nrosched.CheckNROSchedulerAvailable(fxt.Client, nroSchedObj.Name)
+			err = fxt.Client.Delete(context.TODO(), &schedPods[0])
+			Expect(err).ToNot(HaveOccurred())
+			nrosched.CheckNROSchedulerAvailable(fxt.Client, nroSchedObj.Name)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:802]: the scheduler object will still be there what we need to make sure of is that the new pod is restarted and is running

File: test/e2e/serial/tests/configuration.go [hunk 9]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+			By("modifying Topology Manager Policy under kubeletconfig to none")
+			Eventually(func(g Gomega) {
+				err = fxt.Client.Get(context.TODO(), client.ObjectKeyFromObject(targetedKC), targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				kcObj, err = kubeletconfig.MCOKubeletConfToKubeletConf(targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				applyNewTopologyManagerPolicy(&kcObj.TopologyManagerPolicy, v1beta1.NoneTopologyManagerPolicy)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:740]: let's add a comment why to change to "none" specifically or it can be other policy?

File: test/e2e/serial/tests/configuration.go [hunk 10]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			// getting the number of cpus that is required for a numa zone to make a TAE deployment
+			const (
+				NUMAZonesRequired     = 2
+				hostsRequired         = 2
+				cpuResourcePercentage = 1.2 // 1.2 meaning 120 percent
+			)
+			var nrtCandidates []nrtv1alpha2.NodeResourceTopology
+			By(fmt.Sprintf("filtering available nodes with at least %d NUMA zones", NUMAZonesRequired))
+			nrtCandidates = e2enrt.FilterZoneCountEqual(nrtList.Items, NUMAZonesRequired)
+			Expect(len(nrtCandidates)).ToNot(BeNumerically("<", hostsRequired))
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:830]: this can be moved to the very first of the test so if not enough nodes with 2 numas the test should be skipped thus avoid updating the kubeletcofig

File: test/e2e/serial/tests/configuration.go [hunk 11]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+				err = fxt.Client.Update(context.TODO(), targetedKC)
+				g.Expect(err).ToNot(HaveOccurred())
+			}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
+			waitForMcpsCondition(fxt.Client, context.TODO(), mcps, machineconfigv1.MachineConfigPoolUpdated)
+
+			// getting the number of cpus that is required for a numa zone to make a TAE deployment
+			const (
+				NUMAZonesRequired     = 2
+				hostsRequired         = 2
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:824]: for this test one node is also enough

File: test/e2e/serial/tests/configuration.go [hunk 12]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+				cpuResourcePercentage = 1.2 // 1.2 meaning 120 percent
+			)
+			var nrtCandidates []nrtv1alpha2.NodeResourceTopology
+			By(fmt.Sprintf("filtering available nodes with at least %d NUMA zones", NUMAZonesRequired))
+			nrtCandidates = e2enrt.FilterZoneCountEqual(nrtList.Items, NUMAZonesRequired)
+			Expect(len(nrtCandidates)).ToNot(BeNumerically("<", hostsRequired))
+
+			referenceNode := nrtCandidates[0]
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:832]: choosing the first node is fine considering all nodes should be similar in topology let's just add a comment about that to explain the reason behind our decision

File: test/e2e/serial/tests/configuration.go [hunk 13]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+				applyNewTopologyManagerPolicy(&kcObj.TopologyManagerPolicy, v1beta1.SingleNumaNodeTopologyManagerPolicy)
+				err = kubeletconfig.KubeletConfToMCKubeletConf(kcObj, targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				err = fxt.Client.Update(context.TODO(), targetedKC)
+				g.Expect(err).ToNot(HaveOccurred())
+			}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
+			waitForMcpsCondition(fxt.Client, context.TODO(), mcps, machineconfigv1.MachineConfigPoolUpdated)
+
+			// getting the number of cpus that is required for a numa zone to make a TAE deployment
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:821]: let's make this a By()

File: test/e2e/serial/tests/configuration.go [hunk 14]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			referenceNode := nrtCandidates[0]
+			referenceZone := referenceNode.Zones[0]
+			cpuQty, ok := e2enrt.FindResourceAvailableByName(referenceZone.Resources, string(corev1.ResourceCPU))
+			Expect(ok).To(BeTrue(), "no CPU resource in zone %q node %q", referenceZone.Name, referenceNode.Name)
+
+			cpuNum, ok := cpuQty.AsInt64()
+			Expect(ok).To(BeTrue(), "invalid CPU resource in zone %q node %q: %v", referenceZone.Name, referenceNode.Name, cpuQty)
+
+			cpuResources := strconv.Itoa(int(float64(cpuNum) * cpuResourcePercentage))
+
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:841]: lets log out the avialable per numa on the selected node vs what is the cpuResources to be requested in the dp. add that as klog.Infof()

File: test/e2e/serial/tests/configuration.go [hunk 15]
```diff
@@ -693,9 +697,195 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+				time.Sleep(10 * time.Second)
+
+				By(fmt.Sprintf("ensuring the deployment %q keep being pending %d/%d", deployment.Name, step+1, maxStep))
+
+				err = fxt.Client.Get(context.TODO(), client.ObjectKeyFromObject(deployment), updatedDeployment)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(wait.IsDeploymentComplete(deployment, &updatedDeployment.Status)).To(BeFalse(), "deployment %q become ready", deployment.Name)
+			}
+		})
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:857]: the deployment might be unready/pending because several reasons we want to make sure that the reason here is TAE, you can use something like https://github.com/openshift-kni/numaresources-operator/blob/main/test/e2e/serial/tests/workload_unschedulable.go#L308

File: test/e2e/serial/tests/configuration.go (hunk 9 above)
Comment (shajmakh) [reply test/e2e/serial/tests/configuration.go:740]: also let's logout the current policy and if not single-numa-node let's report a warning , and if the same as `none` then break with a klog.Info log

**Review by shajmakh — COMMENTED** [review]
Thanks for this! 
added few comments below for initial review

File: test/e2e/serial/tests/configuration.go (hunk 1 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:701]: okay np

File: test/e2e/serial/tests/configuration.go (hunk 2 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:702]: got it

File: test/e2e/serial/tests/configuration.go (hunk 5 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:751]: yes you're right thanks for that

File: test/e2e/serial/tests/configuration.go (hunk 3 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:732]: ok np

File: test/e2e/serial/tests/configuration.go (hunk 4 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:748]: got it

File: test/e2e/serial/tests/configuration.go (hunk 7 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:798]: yes thanks

File: test/e2e/serial/tests/configuration.go (hunk 1 above)
Comment (ffromani) [inline test/e2e/serial/tests/configuration.go:701]: let's make sure this is `[tagged]` and `Label()`ed `disruptive`, as it restarts the kubelet

Comment (ffromani) [inline test/e2e/serial/tests/configuration.go:701]: please use `It("description", Label(), func(ctx context.Context) { ...`  aka let gingko pass a context, and use it everywhere in the test function body, so we don't need to sprinkle `context.TODO()` everywhere

File: test/e2e/serial/tests/configuration.go (hunk 9 above)
Comment (ffromani) [inline test/e2e/serial/tests/configuration.go:740]: this `applyNewTopologyManagerPolicy` is a one-liner, so i'd drop it and inline the assignment everywhere

File: test/e2e/serial/tests/configuration.go (hunk 10 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:830]: that's right moved it to the top section

File: test/e2e/serial/tests/configuration.go (hunk 11 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:824]: np

Comment (ffromani) [reply test/e2e/serial/tests/configuration.go:824]: this check should be done at the beginning of the test code, to avoid as much as we can any unnecessary churn and keep the test execution time at bay

File: test/e2e/serial/tests/configuration.go (hunk 12 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:832]: np

File: test/e2e/serial/tests/configuration.go (hunk 13 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:821]: got it

**Review by ffromani — COMMENTED** [review]
this is an impressive first contribution, kudos

File: test/e2e/serial/tests/configuration.go (hunk 9 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:740]: It can be any other policy but making it none and then changing to single-numa-node and then creating a tae deployment is the best option in my opinion to check this scenario, but it can be any policy though

File: test/e2e/serial/tests/configuration.go (hunk 1 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:701]: All the tests in this file is under 1 Describe and it already has disruptive should I put another it also under It() or its considered a duplicate?

Comment (ffromani) [reply test/e2e/serial/tests/configuration.go:701]: no, if the parent `Describe` is already tagged and labeled `disruptive` then we're good.

File: test/e2e/serial/tests/configuration.go (hunk 14 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:841]: yes np

File: test/e2e/serial/tests/configuration.go (hunk 9 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:740]: ok got it

File: test/e2e/serial/tests/configuration.go (hunk 6 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:781]: yes fixed

File: test/e2e/serial/tests/configuration.go (hunk 15 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:857]: The topology policy at this point will be `single-numa-node` so the pod will keep on pending if the scheduler got the change correctly so we are expecting here for the pod to be pending because there is not enough cpus on a numa level to schedule the pod and it will fail on failed scheduling rather than TAE so there's no need to check that but there is a need to check for `FailedScheduling`.

Comment (shajmakh) [reply test/e2e/serial/tests/configuration.go:883]: that's right indeed, thanks for pointing this out

File: test/e2e/serial/tests/configuration.go [hunk 16]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			By("getting the number of cpus that is required for a numa zone to make a TAE deployment")
+			const (
+				NUMAZonesRequired     = 2
+				hostsRequired         = 1
+				cpuResourcePercentage = 1.5 // 1.5 meaning 150 percent
+			)
+			var nrtCandidates []nrtv1alpha2.NodeResourceTopology
+			By(fmt.Sprintf("filtering available nodes with at least %d NUMA zones", NUMAZonesRequired))
+			nrtCandidates = e2enrt.FilterZoneCountEqual(nrtList.Items, NUMAZonesRequired)
+			Expect(len(nrtCandidates)).ToNot(BeNumerically("<", hostsRequired))
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:711]: let's make the assertion here a Skip instead of failing. It is okay that the environment is not relevant for a test, we'ld just skip instead of failing, for example:
https://github.com/openshift-kni/numaresources-operator/blob/main/test/e2e/serial/tests/resource_accounting.go#L111

File: test/e2e/serial/tests/configuration.go [hunk 17]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			Eventually(func(g Gomega) {
+				err = fxt.Client.Get(ctx, client.ObjectKeyFromObject(targetedKC), targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				kcObj, err = kubeletconfig.MCOKubeletConfToKubeletConf(targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				// Before restarting the scheduler, we change the topology manager policy to 'none'.
+				// This ensures that if we later change it to 'single NUMA node' and attempt to create a deployment that should fail on TAE,
+				// the deployment will fail as expected. If the policy were still set to 'none', it would allow the deployment to proceed.
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:768]: let's clarify here "..create a deployment that should fail on TAE," that we expect that fail to happen only when the bug occur

File: test/e2e/serial/tests/configuration.go [hunk 18]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+				err = kubeletconfig.KubeletConfToMCKubeletConf(kcObj, targetedKC)
+				Expect(err).ToNot(HaveOccurred())
+
+				err = fxt.Client.Update(ctx, targetedKC)
+				g.Expect(err).ToNot(HaveOccurred())
+			}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
+
+			// TODO find a solution to wait for mcp to start updating only if the tm policy
+			// has changed otherwise it will be stuck here
+			waitForMcpUpdate(fxt.Client, ctx, mcps)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:779]: you can verify that even before by checking if `initialTopologyManagerPolicy != NoneTopologyManagerPolicy` is true only then execute the eventually and wait for the update to happen

File: test/e2e/serial/tests/configuration.go [hunk 19]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+			pod := &pods[0]
+			Expect(pod.Status.Phase).To(Equal(corev1.PodPending))
+
+			schedulerName = nroSchedObj.Status.SchedulerName
+			Expect(schedulerName).ToNot(BeEmpty(), "cannot autodetect the TAS scheduler name from the cluster")
+
+			isFailed, err := nrosched.CheckPODSchedulingFailed(fxt.K8sClient, pod.Namespace, pod.Name, schedulerName)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(isFailed).To(BeFalse(), "pod %s/%s with scheduler %s failed to schedule", pod.Namespace, pod.Name, schedulerName)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:882]: CheckPODSchedulingFailed returns true if the pod indeed failed. if it returned false it either means there was an error or that the pod went running, which then the test should fail: https://github.com/openshift-kni/numaresources-operator/blob/main/test/utils/nrosched/nrosched.go#L88

File: test/e2e/serial/tests/configuration.go [hunk 20]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
 				return kcCmNamesCur
 			}).WithContext(ctx).WithTimeout(5 * time.Minute).WithPolling(10 * time.Second).Should(Equal(kcCmNamesPre))
 		})
+
+		It("[reboot_required][slow][unsched][schedrst] should be able to correctly identify topology manager policy without scheduler restarting", Label("reboot_required", "slow", "unsched", "schedrst"), func(ctx context.Context) {
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:701]: for reboot tests we need to add an additional flag at the beginning of the test to avoid falling on NRT mismatch after reboot. this is a limitation in which kernel might distribute the resources (mainly memory) differently after node reboot. see here for example: https://github.com/openshift-kni/numaresources-operator/blob/main/test/e2e/serial/tests/configuration.go#L109

File: test/e2e/serial/tests/configuration.go (hunk 16 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:711]: yes you're right

File: test/e2e/serial/tests/configuration.go (hunk 17 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:768]: np will do

File: test/e2e/serial/tests/configuration.go (hunk 19 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:882]: got it fixed

File: test/e2e/serial/tests/configuration.go (hunk 20 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:701]: got it ! added it

**Review by Tal-or — COMMENTED** [review]
IIUC this test should verify a behavior that was considered to be a bug and now it was fixed right?
If this is the case can we add a reference to the original bug?

File: test/e2e/serial/tests/configuration.go (hunk 18 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:779]: right thanks!

File: test/e2e/serial/tests/configuration.go [hunk 21]
```diff
@@ -693,9 +697,219 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+				Eventually(func(g Gomega) {
+					err = kubeletconfig.KubeletConfToMCKubeletConf(kcObj, targetedKC)
+					Expect(err).ToNot(HaveOccurred())
+
+					// changing the topology manager policy back to the initial state when the test has finished or failed
+					kcObj.TopologyManagerPolicy = initialTopologyManagerPolicy
+
+					err = fxt.Client.Update(ctx, targetedKC)
+					g.Expect(err).ToNot(HaveOccurred())
+				}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
```
[the remark this answers is not in the document]
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:807]: understood and addressed!

File: test/e2e/serial/tests/configuration.go [hunk 22]
```diff
@@ -693,9 +697,216 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+						kcObj, err = kubeletconfig.MCOKubeletConfToKubeletConf(targetedKC)
+						Expect(err).ToNot(HaveOccurred())
+
+						err = kubeletconfig.KubeletConfToMCKubeletConf(kcObj, targetedKC)
+						Expect(err).ToNot(HaveOccurred())
+
+						// changing the topology manager policy back to the initial state when the test has finished or failed
+						kcObj.TopologyManagerPolicy = initialTopologyManagerPolicy
+
+						err = fxt.Client.Update(ctx, targetedKC)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:801]: this code is making the required change (`kcObj.TopologyManagerPolicy = initialTopologyManagerPolicy`) after converting the kcObj to targetedKC which is not going to make any difference at all in the targetedKC, lines 798-799 should be switched with 792-793, same for similar instances

Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:801]: you're right! thanks!

File: test/e2e/serial/tests/configuration.go [hunk 23]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			schedPods, err := podlist.With(fxt.Client).ByDeployment(ctx, *schedDeployment)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(len(schedPods)).To(Equal(1))
+
+			schedulerName = nroSchedObj.Status.SchedulerName
+			Expect(schedulerName).ToNot(BeEmpty(), "cannot autodetect the TAS scheduler name from the cluster")
+
+			// deleting the NRO scheduler
+			By(fmt.Sprintf("deleting the NRO Scheduler object: %s", nroSchedObj.Name))
```
[the remark this answers is not in the document]
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:827]: okay np

File: test/e2e/serial/tests/configuration.go [hunk 24]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+	podLabels := map[string]string{
+		"test": "test-dp",
+	}
+	nodeSelector := map[string]string{}
+	deployment := objects.NewTestDeployment(replicas, podLabels, nodeSelector, fxt.Namespace.Name, name, images.GetPauseImage(), []string{images.PauseCommand}, []string{})
+	deployment.Spec.Template.Spec.SchedulerName = schedulerName
+	deployment.Spec.Template.Spec.Containers[0].Resources.Limits = corev1.ResourceList{
+		corev1.ResourceCPU:    resource.MustParse(cpus),
+		corev1.ResourceMemory: resource.MustParse("16Gi"),
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:899]: from the run logs looks like this is the reason the pod is unschedualable:

```

I0729 11:01:06.244631 3703411 nrosched.go:84] -> found relevant scheduling event for pod e2e-test-configuration-7485/testdp-85f4c8ccf7-v6v4h: [type="Warning" action="Scheduling" message="0/5 nodes are available: 2 Insufficient memory, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/5 nodes are available: 2 No preemption victims found for incoming pod, 3 Preemption is not helpful for scheduling."
```
let's adjust this to 256Mi

File: test/e2e/serial/tests/configuration.go [hunk 25]
```diff
@@ -693,9 +697,215 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(pods)).To(Equal(1))
+
+			pod := &pods[0]
+			Expect(pod.Status.Phase).To(Equal(corev1.PodPending))
+
+			schedulerName = nroSchedObj.Status.SchedulerName
+			Expect(schedulerName).ToNot(BeEmpty(), "cannot autodetect the TAS scheduler name from the cluster")
+
+			isFailed, err := nrosched.CheckPODSchedulingFailed(fxt.K8sClient, pod.Namespace, pod.Name, schedulerName)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:880]: to avoid false negative we can replace this general failure function with something like https://github.com/openshift-kni/numaresources-operator/blob/main/test/e2e/serial/tests/resource_hostlevel.go#L340 but using the right message about insufficient cpus, you can check that from the failing pod's events

File: test/e2e/serial/tests/configuration.go (hunk 24 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:899]: Got it! Changed.

File: test/e2e/serial/tests/configuration.go (hunk 25 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:880]: got it np, will change that

File: test/e2e/serial/tests/configuration.go [hunk 26]
```diff
@@ -693,9 +697,220 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+
+					err = fxt.Client.Update(ctx, targetedKC)
+					g.Expect(err).ToNot(HaveOccurred())
+				}).WithTimeout(10 * time.Minute).WithPolling(30 * time.Second).Should(Succeed())
+
+				waitForMcpUpdate(fxt.Client, ctx, mcps)
+			}
+
+			defer func() {
+				err = fxt.Client.Get(ctx, client.ObjectKeyFromObject(targetedKC), targetedKC)
```
Comment (shajmakh) [inline test/e2e/serial/tests/configuration.go:787]: let's use local vars as much as possible in the implemented functions of defer and eventually like err and kcObj, for example the first initialization of each is associated with := or

**Comment by rshemtov13** [comment]
after PR is merged will add test_id

File: test/e2e/serial/tests/configuration.go (hunk 8 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go]: Has been changed thank you!

**Comment by shajmakh** [comment]
Thank you for addressing all the comments and for all the efforts on this!
/lgtm
/approve
/hold
for @ffromani 's ack or others review

File: test/e2e/serial/tests/configuration.go [hunk 27]
```diff
@@ -693,9 +697,222 @@ var _ = Describe("[serial][disruptive] numaresources configuration management",
...
+			// choosing the first node because the nodes should be similar in topology so it doesnt matter which one we choose
+			referenceNode := nrtCandidates[0]
+			referenceZone := referenceNode.Zones[0]
+
+			cpuQty, ok := e2enrt.FindResourceAvailableByName(referenceZone.Resources, string(corev1.ResourceCPU))
+			Expect(ok).To(BeTrue(), "no CPU resource in zone %q node %q", referenceZone.Name, referenceNode.Name)
+
+			cpuNum, ok := cpuQty.AsInt64()
+			Expect(ok).To(BeTrue(), "invalid CPU resource in zone %q node %q: %v", referenceZone.Name, referenceNode.Name, cpuQty)
+			klog.Infof("available CPUs per numa on the node: %d, ", cpuNum)
```
Comment (ffromani) [inline test/e2e/serial/tests/configuration.go:726]: nit: let's use https://kubernetes.io/blog/2020/09/04/kubernetes-1-19-introducing-structured-logs/ in new code (Infof -> InfoS)

**Review by ffromani — COMMENTED** [review]
/lgtm
/unhold

thanks, very nice first contribution!

File: test/e2e/serial/tests/configuration.go (hunk 27 above)
Comment (rshemtov13) [reply test/e2e/serial/tests/configuration.go:726]: np will do!
